encoder.go

```diff
@@ -228,45 +270,41 @@ func (encoder *encoder) encodeStruct(value reflect.Value, obj *wafObject, depth
 // - Even if the element values are invalid or null we still keep them to report the map key
 func (encoder *encoder) encodeMap(value reflect.Value, obj *wafObject, depth int) {
 	capacity := value.Len()
-	length := 0
 	if capacity > encoder.containerMaxSize {
 		capacity = encoder.containerMaxSize
 	}
 
 	objArray := encoder.cgoRefs.AllocWafArray(obj, wafMapType, uint64(capacity))
+
+	length := 0
 	for iter := value.MapRange(); iter.Next(); {
 		if length == capacity {
+			encoder.addTruncation(ContainerTooLarge, value.Len())
 			break
 		}
 
 		objElem := &objArray[length]
-		if encoder.encodeMapKey(iter.Key(), objElem, depth) != nil {
+		if err := encoder.encodeMapKeyFromValue(iter.Key(), objElem); err != nil {
 			continue
 		}
 
 		if err := encoder.encode(iter.Value(), objElem, depth); err != nil {
 			// We still need to keep the map key, so we can't discard the full object, instead, we make the value a noop
 			encodeNative[uintptr](0, wafInvalidType, objElem)
 		}
+
 		length++
 	}
 
 	// Fix the size because we skipped map entries
 	obj.nbEntries = uint64(length)
 }
 
-// encodeMapKey takes a reflect.Value and a wafObject and returns a wafObject ready to be considered a map key
+// encodeMapKeyFromValue takes a reflect.Value and a wafObject and returns a wafObject ready to be considered a map key
 // We use the function cgoRefPool.AllocWafMapKey to store the key in the wafObject. But first we need
 // to grab the real underlying value by recursing through the pointer and interface values.
-func (encoder *encoder) encodeMapKey(value reflect.Value, obj *wafObject, depth int) error {
-	kind := value.Kind()
-	for ; depth > 0 && (kind == reflect.Pointer || kind == reflect.Interface); value, kind = value.Elem(), value.Elem().Kind() {
-		if value.IsNil() {
-			return errInvalidMapKey
-		}
-
-		depth--
-	}
+func (encoder *encoder) encodeMapKeyFromValue(value reflect.Value, obj *wafObject) error {
```

Comment:
`EncodeMapKey` and `EncodeMapKeyFromValue` names are not very useful to differentiate them. How about merging them into one or renaming `EncoreMapKey` to `EncodeMapKeyFromString`